### PR TITLE
add __hash__ to FunctionSchema

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90735
* #90734
* #90733
* #90732
* #90731
* __->__ #90730

This PR adds __hash__ to FunctionSchema pybind binding, so that
it could be used for things like dict indexing

### CODE REVIEW

File: torch/csrc/jit/python/init.cpp [hunk 1]
```diff
@@ -1700,6 +1700,13 @@ void initJITBindings(PyObject* module) {
           [](const FunctionSchema& self, const FunctionSchema& other) {
             return self == other;
           })
+      .def(
+          "__hash__",
+          [](const FunctionSchema& self) {
+            std::stringstream ss;
+            ss << self;
+            return std::hash<std::string>()(ss.str());
```
Comment (ezyang) [inline torch/csrc/jit/python/init.cpp:1708]: This is a very inefficient way to go about hashing. More conventional way to do hashing is to inspect the equality function, hash each component and combine the hashes

**Review by ezyang — CHANGES_REQUESTED** [review]

File: torch/csrc/jit/python/init.cpp (hunk 1 above)
Comment (wanchaol) [reply torch/csrc/jit/python/init.cpp]: Thanks for the guide! Just updated the PR to use individual component hashing and combine them.

File: aten/src/ATen/core/function_schema.h [hunk 2]
```diff
@@ -639,6 +639,40 @@ template<>
       return c10::hash_combine(std::hash<size_t>()(arg.index), std::hash<size_t>()(static_cast<std::size_t>(arg.type)));
     }
   };
+template<>
+  struct hash<c10::Argument> {
+    size_t operator()(const c10::Argument& arg) const
+    {
+      // We don't need to hash default value and alias_info
+      // because they are not useful in distinguishing schema.
```
Comment (ezyang) [inline aten/src/ATen/core/function_schema.h:647]: The hash algorithm should respect equality?

Comment (ezyang) [reply aten/src/ATen/core/function_schema.h:647]: Like, if you have FunctionSchemas with differing default values hash to the same thing, then why did you just hash on the name of the schema only. Would be a lot simpler!

Comment (wanchaol) [reply aten/src/ATen/core/function_schema.h:647]: I see, yeah so far I don't think we have two schemas with exact same arguments and only differs in default values, and couldn't reason about having such ones in the future, that was the primary reason I chose to not hash default value. But if we want to be more respectful to equality, I can surely add the default value hash. I think we still want to leave out the alias_info as it contains a ton of information which is difficult to hash and not really useful in distinguishing schemas. 

I'll update the PR to include default value hashing.

File: aten/src/ATen/core/function_schema.h [hunk 3]
```diff
@@ -639,6 +639,44 @@ template<>
...
+      if (arg.default_value()) {
+        auto default_value_hash = c10::hash<c10::IValue>{}(arg.default_value().value());
+        hash = c10::hash_combine(hash, default_value_hash);
+      }
+      if (arg.N()) {
+        auto N_hash = std::hash<int64_t>{}(*arg.N());
+        hash = c10::hash_combine(hash, N_hash);
+      }
+      // We don't need to hash the alias_info because it
+      // is not useful in distinguishing schema.
```
Comment (ezyang) [inline aten/src/ATen/core/function_schema.h:660]: alias_info is tested in equality. Please also hash it.

File: aten/src/ATen/core/alias_info.h [hunk 4]
```diff
@@ -117,3 +118,25 @@ inline std::ostream& operator<<(std::ostream& out, const AliasInfo& aliasInfo) {
...
 } // namespace c10
+
+namespace std {
+template <>
+  struct hash<c10::AliasInfo> {
+    size_t operator()(const c10::AliasInfo& aliasInfo) const {
+      auto hash = std::hash<bool>()(aliasInfo.isWrite());
+      for (auto &e: aliasInfo.beforeSets()) {
+        auto symbol_hash = std::hash<c10::Symbol>()(e);
+        hash = c10::hash_combine(hash, symbol_hash);
```
Comment (ezyang) [inline aten/src/ATen/core/alias_info.h:129]: This is not correct. Iteration order over unordered set is not specified. Two unordered sets can have the same contents but differing iteration order. There are a few ways to solve this problem but a cheap and cheerful one is to use a commutative operator like addition or xor to combine the hashes from the operations in the set.

Comment (wanchaol) [reply aten/src/ATen/core/alias_info.h:129]: Got it, thanks! Let me try this out.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**